Select the integration sector added

Mukul1235 opened the pull request:
<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->
issue link: https://github.com/zulip/zulip/issues/30139

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

Mukul1235 commented:
@zulipbot  add maintainer review

Mukul1235 commented:
@zulipbot add "maintainer review"

alya commented:
Please clean up your commit history and post again to request a review. See [here](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for guidelines.

Mukul1235 commented:
@zulipbot add "maintainer review"

alya commented:
Please update your commit message to match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages).

You will also need to complete the self-review checklist. You might want to go through our[guide on submitting a pull request](https://zulip.readthedocs.io/en/latest/contributing/reviewable-prs.html) to learn how to present your proposed changes to Zulip, so that your work can be reviewed.

Mukul1235 commented:
@zulipbot add "maintainer review"

Mukul1235 commented:
> Please update your commit message to match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages).
> 
> You will also need to complete the self-review checklist. You might want to go through our [guide on submitting a pull request](https://zulip.readthedocs.io/en/latest/contributing/reviewable-prs.html) to learn how to present your proposed changes to Zulip, so that your work can be reviewed.


hi @alya 
Could you please review it once

alya commented:
@PieterCK would you be able to do an initial review pass here?

@Mukul1235 Your commit titles are missing periods, so please fix that up as well.

PieterCK commented:
@Mukul1235 - Thanks for working on this issue! :smiley: 

To get a better understanding of the pull request review process in Zulip, you can check out our [documentation on the stages of a pull request](https://zulip.readthedocs.io/en/latest/contributing/review-process.html). We use labels to track the different stages of the review, and our maintainers and reviewers will generally be the ones to update these labels to track where your PR is in the process. CMIIW, @sbansal1999 

@zulipbot remove "maintainer review"

PieterCK commented:
@Mukul1235 - I think you need to make sure your commits to follow our [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html). I'll point out a couple of things to consider

1. Focus on the **why** and **how** of the change instead of the **what**s.
>The body of the commit message should explain why and how the change was made. Like a good code comment, it should provide context and motivation that will help both a reviewer now, and a developer looking at your changes a year later, understand the motivation behind your decisions.
2. I see that in the second commit, you described a couple of changes, are there any changes that you can split down to its own commit? generally, we want to keep each commit small and self-contained. 
https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commits-should-generally-be-minimal
3. You can use our [tool](https://zulip.readthedocs.io/en/latest/git/zulip-tools.html) to automatically lint your commit description and summary format.

Additionally, it would be helpful to tidy up the PR description. Try to explain in more detail why and how you implemented it the way you did.

Writing clear explanations for your changes not only helps you get reviews faster but also helps you catch bugs and improve as a developer. I'll continue to review the code once you've reorganized the commits. Let me know if you have any concerns or questions!

Mukul1235 commented:
> @Mukul1235 - I think you need to make sure your commits to follow our [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html). I'll point out a couple of things to consider
> 
> 1. Focus on the **why** and **how** of the change instead of the **what**s.
> 
> > The body of the commit message should explain why and how the change was made. Like a good code comment, it should provide context and motivation that will help both a reviewer now, and a developer looking at your changes a year later, understand the motivation behind your decisions.
> 
> 2. I see that in the second commit, you described a couple of changes, are there any changes that you can split down to its own commit? generally, we want to keep each commit small and self-contained.
>    https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commits-should-generally-be-minimal
> 3. You can use our [tool](https://zulip.readthedocs.io/en/latest/git/zulip-tools.html) to automatically lint your commit description and summary format.
> 
> Additionally, it would be helpful to tidy up the PR description. Try to explain in more detail why and how you implemented it the way you did.
> 
> I'll continue to review the code once you've reorganized the commits. Let me know if you have any concerns or questions!

@PieterCK Actually In the second Commit I fixed all the Tests That were failing. So If I Separate the Tests then Commit might not be Coherent

PieterCK reviewed (commented):
@Mukul1235 - Hi, I think we can move and consolidate some changes from the second commit to this commit. We can make this commit only deal with adding `integration_name` field to the bot creation modal and saving it to BotConfigData. 

with that in mind, the commit description and summary for that change may look something like:

> bots settings: Add `integration_name` to incoming webhook bot.
> 
> In #30139, we're planning to automatically populate the avatar for 
incoming webhook bots. To that end, this commit makes it possible 
to associate an incoming webhook bot with an integration by adding 
a "Select an integration" dropdown field to the bot creation modal and 
saving it as integration_name in the incoming webhook bots' BotConfigData.
> 
> Fixes part of #30139.

PieterCK commented on zerver/views/users.py:
```diff
@@ -115,8 +115,9 @@ def deactivate_user_backend(
     user_id: PathOnly[int],
-    deactivation_notification_comment: Annotated[str, StringConstraints(max_length=2000)]
-    | None = None,
+    deactivation_notification_comment: (
```
I think you can move all the changes in this file to the first commit. The changes in the first commit would group together nicely up to this point.

If you need to add more changes here that are unrelated to the first commit, you can add a `# TODO` comment in that commit describing what you plan to do next. Then, in the next commit, you can replace that comment with the actual changes.

PieterCK commented on zerver/views/users.py:
```diff
@@ -452,7 +453,7 @@ def add_bot_backend(
     bot_type: Json[int] = UserProfile.DEFAULT_BOT,
     payload_url: Json[Annotated[str, AfterValidator(check_url)]] = "",
     service_name: str | None = None,
-    integration_name: Optional[str] = None,
```
Please fix new changes in the same commit it is introduced, and not in the following commit.

Mukul1235 replied:
Changes made

PieterCK reviewed (commented):
I think we should first focus on improving the initial commit by adding the integration_name field to the bot creation modal and ensuring it is saved correctly to BotConfigData. After that, we'll continue working on the rest of the PR. I expect to continue reviewing this PR in the coming days. Thank you!

PieterCK commented:
We should make the Integration Name field appear only when the user selects the "Incoming webhook" bot type. 
![image](https://github.com/user-attachments/assets/ee817b53-ee44-4b74-863b-cfc998985134)

Mukul1235 commented:
> We only need to show the user the "Integration name" dropdown if the selected bot type is an Incoming webhook bot. This should be included in the first commit.

Currently, I have added "Integration name" to every Bot type. Do we only need the integration_name for Incoming Webhook?

Mukul1235 commented on zerver/views/users.py:
```diff
@@ -561,7 +562,8 @@ def add_bot_backend(
         )
-
+    integration_name = integration_name or ""
+    set_bot_config(bot_profile, "integration_name", integration_name)
```
conditions added checking- integration_name is not None and bot_type===INCOMING_WEBHOOK_BOT

Mukul1235 commented on web/templates/settings/add_new_bot_form.hbs:
```diff
@@ -13,6 +13,17 @@
         </div>
+        <div class="input-group">
+            <label for="integration_type" class="modal-field-label">
+                {{t "Integration Type" }}
```
Changes made

Mukul1235 commented:
Hi @PieterCK 
Did you got the time to review this PR?

PieterCK commented on zerver/views/users.py:
```diff
@@ -115,8 +115,9 @@ def deactivate_user_backend(
-    | None = None,
+    deactivation_notification_comment: (
+        Annotated[str, StringConstraints(max_length=2000)] | None
+    ) = None,
```
Hmm why is this line changed? 
Have you fetch upstream and rebase upstream/main ?

PieterCK commented on zerver/views/users.py:
```diff
@@ -529,7 +531,6 @@ def add_bot_backend(
         (default_events_register_stream, ignored_sub) = access_stream_by_name(
             user_profile, default_events_register_stream_name
         )
-
```
We generally don't change anything unrelated to the issue at hand.

Mukul1235 replied:
Removed the change

PieterCK commented on web/src/settings_bots.js:
```diff
@@ -221,17 +241,30 @@ export function add_a_new_bot() {
+                break;
+            }
+            case INCOMING_WEBHOOK_BOT_TYPE: {
+                // console.log("Bot_type Incoming Webhook");
```
I think you forgot to clean up this

Mukul1235 replied:
Changes made

PieterCK commented on zerver/actions/users.py:
```diff
@@ -554,7 +554,9 @@ def do_update_outgoing_webhook_service(
                 user_id=bot_profile.id,
                 services=[
                     dict(
-                        base_url=service.base_url, interface=service.interface, token=service.token
```
Once again, we usually don't reformat anything that's irrelevant to the overall issue of this PR.

Mukul1235 replied:
Fixed

PieterCK commented on zerver/actions/users.py:
```diff
@@ -604,6 +606,12 @@ def get_service_dicts_for_bot(user_profile_id: int) -> list[dict[str, Any]]:
+    elif user_profile.bot_type == UserProfile.INCOMING_WEBHOOK_BOT:
+        return [
+            {
+                "integration_name": get_bot_config(user_profile).get("integration_name"),
```
We can return the data structure similar to embedded bot's just above this condition to make things a bit more consistent.
```
try:
    return [
        {
            "config_data": get_bot_config(user_profile),
        }
    ]
except ConfigError:
    return []
```
Don't forget to except `ConfigError`, since the "integration_name" field is not mandatory and can be left empty

Mukul1235 replied:
Changes made

PieterCK commented on zerver/actions/users.py:
The home page crashes when I rebase to the first commit. I think we can move this to its own commit or the next one since it doesn't group together well with the change in the first commit. 

I guess this change is made so that we can fetch "integration_name" in the fronted right? I think this is outside of the scope of the first commit (as per the commit description). 

![image](https://github.com/user-attachments/assets/e72647c8-899c-4140-bfef-22b30e868637)

PieterCK replied:
So referring back to the points about the first commit that you described in the PR description, I think we can leave out the third point about fetching integration_bot configs in the front end. So this is what we need to focus on in the first commit:
- settings_bot.js : get_options_for_integration_input_dropdown_widget method fetch all the integration_type available along with this sending to the backend in formdata.
- add_new_bot_form.hbs: add integration_type in UI.
- users.py: Saving the integration_type in the BotConfigData.

Mukul1235 replied:
Yes, the changes in this file were to fetch "integration_name" in the frontend. Pushing these changes to 2nd commit in which we require fetching of "integration_name" to prepopulate this field.

PieterCK commented:
Hmm the "Generate URL for integration" icon / modal is gone.
![image](https://github.com/user-attachments/assets/0eb609e4-296c-4081-8fe1-196205315b3e)

PieterCK commented:
@Mukul1235 Hi. sorry for the delay. I've dropped some more reviews around the first commit, I feel like we're starting to move in the right direction. Let's continue polishing the first commit. 

If you run `./tools/lint` you can see that your commit description still doesn't follow our commit discipline, you can read the doc again if you're unsure.
![image](https://github.com/user-attachments/assets/990acb14-6759-400f-8c46-fd5bf53753b5)

Mukul1235 commented:
> @Mukul1235 Hi. sorry for the delay. I've dropped some more reviews around the first commit, I feel like we're starting to move in the right direction. Let's continue polishing the first commit.
> 
> If you run `./tools/lint` you can see that your commit description still doesn't follow our commit discipline, you can read the doc again if you're unsure. ![image](https://private-user-images.githubusercontent.com/108744694/358242979-990acb14-6759-400f-8c46-fd5bf53753b5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM4MDU2NzIsIm5iZiI6MTcyMzgwNTM3MiwicGF0aCI6Ii8xMDg3NDQ2OTQvMzU4MjQyOTc5LTk5MGFjYjE0LTY3NTktNDAwZi04YzQ2LWZkNWJmNTM3NTNiNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxNlQxMDQ5MzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02ZTQ2ZDA5YzJjMmM3NGU2YTdlMWNlYmZmNjY0MDk0MGVmYzE4NDI1NGY2MzEwOTI2NjJiMWY1MzQxNjAxMzU3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.UmEgZcTI9-R2vqgOPFlcMQHsByKb_PE454Xjty0iZ34)
> 
> Let me know if you have any doubts

Fixed

Mukul1235 commented:
> > @Mukul1235 Hi. sorry for the delay. I've dropped some more reviews around the first commit, I feel like we're starting to move in the right direction. Let's continue polishing the first commit.
> > If you run `./tools/lint` you can see that your commit description still doesn't follow our commit discipline, you can read the doc again if you're unsure. ![image](https://private-user-images.githubusercontent.com/108744694/358242979-990acb14-6759-400f-8c46-fd5bf53753b5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM4MDU2NzIsIm5iZiI6MTcyMzgwNTM3MiwicGF0aCI6Ii8xMDg3NDQ2OTQvMzU4MjQyOTc5LTk5MGFjYjE0LTY3NTktNDAwZi04YzQ2LWZkNWJmNTM3NTNiNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxNlQxMDQ5MzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02ZTQ2ZDA5YzJjMmM3NGU2YTdlMWNlYmZmNjY0MDk0MGVmYzE4NDI1NGY2MzEwOTI2NjJiMWY1MzQxNjAxMzU3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.UmEgZcTI9-R2vqgOPFlcMQHsByKb_PE454Xjty0iZ34)
> > Let me know if you have any doubts
> 
>Hi @PieterCK no problem. I have fixed the changes for the first commit

PieterCK commented on zerver/tests/test_bots.py:
```diff
@@ -1788,6 +1788,24 @@ def test_create_embedded_bot(self, **extras: Any) -> None:
+    def test_get_bot_config_exception(self, **extras: Any) -> None:
+        self.user_profile = self.example_user("hamlet")
+        self.login_user(self.user_profile)
+        # Bot information
```
we can drop this comment

Mukul1235 replied:
changes made!

PieterCK commented:
@Mukul1235 I dropped a couple more reviews. Let's add a validator that checks for "integration_name" before we save it to `BotConfigData` we want to make sure that its value must be in the `WEBHOOK_INTEGRATIONS` list. we're already doing a similar thing with Embedded bot:
```
# view/user.py/add_bot_backend()
...
if service_name not in [bot.name for bot in EMBEDDED_BOTS]:
      raise JsonableError(_("Invalid embedded bot name."))
```

Other than that, the "Generate URL for an integration" icon/modal is still missing.
![image](https://github.com/user-attachments/assets/99bc7bbc-4ed5-49c1-a071-c2a177a870e5)

Mukul1235 commented on web/src/settings_bots.js:
```diff
@@ -198,10 +199,28 @@ export function update_bot_permissions_ui() {
+    function get_options_for_integration_input_dropdown_widget() {
+        const options = [
+            default_integration_option,
+            ...realm.realm_incoming_webhook_bots
```
changes made!

Mukul1235 commented on zerver/tests/test_bots.py:
```diff
@@ -1891,11 +1909,13 @@ def test_create_incoming_webhook_bot_without_service_name(self) -> None:
             "full_name": "My Stripe Bot",
             "short_name": "my-stripe",
             "bot_type": UserProfile.INCOMING_WEBHOOK_BOT,
+            "integration_name": "Test",
```
changes made!

timabbott commented:
Closing in favor of https://github.com/zulip/zulip/pull/33441, which continues this. Thanks for the effort @Mukul1235!
